File: abc122/d.py
import functools
import re

MOD = 10 ** 9 + 7
ACGT = 'ACGT'
GOOD = 'TTT'
BAD = re.compile(r'AGC|GAC|ACG|A.GC|AG.C')


@functools.lru_cache(maxsize=None)
def dfs(n, i, last3):
    if i == n:
        return 1
    else:
        return sum(
            dfs(n, i + 1, last3[1:] + c)
            for c in ACGT
            if not re.search(BAD, last3 + c)) % MOD


N = int(input())
print(dfs(N, 0, GOOD))

File: abc122/d.rb
# frozen_string_literal: true

BAD = /AGC|GAC|ACG|A.GC|AG.C/.freeze
MOD = 10**9 + 7
CHARS = 'ACGT'.each_char

# memoized DFS
def dfs(i, last3, n, memo)
  return 1 if i == n

  # p [i, last3, memo]
  m = memo[i][last3]
  return m if m

  a =
    CHARS
    .reject { |c| last3 + c =~ BAD }
    .sum { |c| dfs(i + 1, last3[1..-1] + c, n, memo) }

  memo[i][last3] = a % MOD
end

n = gets.to_i
p dfs(0, 'TTT', n, Array.new(n) { {} })
